diff --git a/nginx.spec b/nginx.spec
index decf820..b67e3d4 100644
--- a/nginx.spec
+++ b/nginx.spec
@@ -68,12 +68,21 @@ Source9: nginx.upgrade.sh
 Source10: nginx.suse.logrotate
 Source11: nginx-debug.service
 Source12: COPYRIGHT
+Source100: headers-more-nginx-module
+Source101: ngx-fancyindex
+Source102: nginx_ajp_module
+Source103: nginx-xslt-html-parser.patch
+Source104: nginx-auth-ldap
+Source105: nginx-http-shibboleth
 
 License: 2-clause BSD-like license
 
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
 BuildRequires: zlib-devel
 BuildRequires: pcre-devel
+BuildRequires: openldap-devel
+BuildRequires: libxml2-devel
+BuildRequires: libxslt-devel
 
 Provides: webserver
 
@@ -92,6 +101,12 @@ sed -e 's|%%DEFAULTSTART%%|2 3 4 5|g' -e 's|%%DEFAULTSTOP%%|0 1 6|g' \
     -e 's|%%PROVIDES%%|nginx|g' < %{SOURCE2} > nginx.init
 sed -e 's|%%DEFAULTSTART%%||g' -e 's|%%DEFAULTSTOP%%|0 1 2 3 4 5 6|g' \
     -e 's|%%PROVIDES%%|nginx-debug|g' < %{SOURCE2} > nginx-debug.init
+cp -R -p %SOURCE100 .
+cp -R -p %SOURCE101 .
+cp -R -p %SOURCE102 .
+patch -p1 < %SOURCE103
+cp -R -p %SOURCE104 .
+cp -R -p %SOURCE105 .
 
 %build
 ./configure \
@@ -134,6 +149,12 @@ sed -e 's|%%DEFAULTSTART%%||g' -e 's|%%DEFAULTSTOP%%|0 1 2 3 4 5 6|g' \
         %{?with_http2:--with-http_v2_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
         --with-debug \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 %{__mv} %{_builddir}/%{name}-%{version}/objs/nginx \
@@ -177,6 +198,12 @@ make %{?_smp_mflags}
         --with-ipv6 \
         %{?with_http2:--with-http_v2_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 
