diff --git a/nginx.spec b/nginx.spec
index 281e40b..8a9a7d2 100644
--- a/nginx.spec
+++ b/nginx.spec
@@ -59,7 +59,7 @@ Requires: systemd
 Summary: High performance web server
 Name: nginx
 Version: 1.8.0
-Release: 1%{?dist}.ngx
+Release: 2%{?dist}.ngx
 Vendor: nginx inc.
 URL: http://nginx.org/
 
@@ -74,12 +74,21 @@ Source7: nginx.suse.init
 Source8: nginx.service
 Source9: nginx.upgrade.sh
 Source10: nginx.suse.logrotate
+Source100: headers-more-nginx-module
+Source101: ngx-fancyindex
+Source102: nginx_ajp_module
+Source103: nginx-xslt-html-parser.patch
+Source104: nginx-auth-ldap
+Source105: nginx-http-shibboleth
 
 License: 2-clause BSD-like license
 
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
 BuildRequires: zlib-devel
 BuildRequires: pcre-devel
+BuildRequires: openldap-devel
+BuildRequires: libxml2-devel
+BuildRequires: libxslt-devel
 
 Provides: webserver
 
@@ -100,6 +109,12 @@ Not stripped version of nginx built with the debugging log support.
 
 %prep
 %setup -q
+cp -R -p %SOURCE100 .
+cp -R -p %SOURCE101 .
+cp -R -p %SOURCE102 .
+patch -p1 < %SOURCE103
+cp -R -p %SOURCE104 .
+cp -R -p %SOURCE105 .
 
 %build
 ./configure \
@@ -137,6 +152,13 @@ Not stripped version of nginx built with the debugging log support.
         --with-debug \
         %{?with_spdy:--with-http_spdy_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
+        --with-http_spdy_module \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 %{__mv} %{_builddir}/%{name}-%{version}/objs/nginx \
@@ -175,6 +197,13 @@ make %{?_smp_mflags}
         --with-ipv6 \
         %{?with_spdy:--with-http_spdy_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
+        --with-http_spdy_module \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 
