diff --git a/nginx.spec b/nginx.spec
index e829ec7..8d4aebb 100644
--- a/nginx.spec
+++ b/nginx.spec
@@ -44,10 +44,10 @@ Requires(pre): pwdutils
 
 # end of distribution specific definitions
 
-Summary: High performance web server
+Summary: High performance web server (Built with LDAP, XSLT, XSLT HTML Parser, and Shibboleth module)
 Name: nginx
 Version: 1.6.2
-Release: 1%{?dist}.ngx
+Release: 2%{?dist}.ngx
 Vendor: nginx inc.
 URL: http://nginx.org/
 
@@ -61,12 +61,21 @@ Source6: nginx.vh.example_ssl.conf
 Source7: nginx.suse.init
 Source8: nginx.service
 Source9: nginx.upgrade.sh
+Source10: headers-more-nginx-module
+Source11: ngx-fancyindex
+Source12: nginx_ajp_module
+Source13: nginx-xslt-html-parser.patch
+Source14: nginx-auth-ldap
+Source15: nginx-http-shibboleth
 
 License: 2-clause BSD-like license
 
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
 BuildRequires: zlib-devel
 BuildRequires: pcre-devel
+BuildRequires: openldap-devel
+BuildRequires: libxml2-devel
+BuildRequires: libxslt-devel
 
 Provides: webserver
 
@@ -83,6 +92,12 @@ Not stripped version of nginx built with the debugging log support.
 
 %prep
 %setup -q
+cp -R -p %SOURCE10 .
+cp -R -p %SOURCE11 .
+cp -R -p %SOURCE12 .
+patch -p1 < %SOURCE13
+cp -R -p %SOURCE14 .
+cp -R -p %SOURCE15 .
 
 %build
 ./configure \
@@ -120,6 +135,13 @@ Not stripped version of nginx built with the debugging log support.
         --with-debug \
         %{?with_spdy:--with-http_spdy_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
+        --with-http_spdy_module \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 %{__mv} %{_builddir}/%{name}-%{version}/objs/nginx \
@@ -158,6 +180,13 @@ make %{?_smp_mflags}
         --with-ipv6 \
         %{?with_spdy:--with-http_spdy_module} \
         --with-cc-opt="%{optflags} $(pcre-config --cflags)" \
+        --with-http_spdy_module \
+        --with-http_xslt_module \
+        --add-module=%{_builddir}/%{name}-%{version}/ngx-fancyindex \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx_ajp_module \
+        --add-module=%{_builddir}/%{name}-%{version}/headers-more-nginx-module \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-auth-ldap \
+        --add-module=%{_builddir}/%{name}-%{version}/nginx-http-shibboleth \
         $*
 make %{?_smp_mflags}
 
